### PR TITLE
provider cmds: storage-ask and retrieval-ask

### PR DESCRIPTION
**Pull request by nonsense**
Fixes: https://github.com/filecoin-project/boost/issues/449

TODO:
- [x] storage ask over libp2p
- [x] retrieval ask over libp2p

### CODE REVIEW

File: cmd/boost/provider_cmd.go [hunk 1]
```diff
@@ -97,3 +103,150 @@ var libp2pInfoCmd = &cli.Command{
...
+			afmt.Println("Usage: query-ask [minerAddress]")
+			return nil
+		}
+
+		maddr, err := address.NewFromString(cctx.Args().First())
+		if err != nil {
+			return err
+		}
+
+		api, closer, err := lcli.GetFullNodeAPI(cctx)
```
Comment (nonsense) [inline cmd/boost/provider_cmd.go:137]: Note that unlike other `boost` cmds, this one expects a Full Node, not just a Gateway.

Comment (nonsense) [reply cmd/boost/provider_cmd.go]: @dirkmc this is no longer the case, we query for asks directly over libp2p and just use the gateway to fetch multiaddrs for the miner.

File: cmd/boost/provider_cmd.go [hunk 2]
```diff
@@ -97,3 +109,192 @@ var libp2pInfoCmd = &cli.Command{
 		return nil
 	},
 }
+
+var queryAskCmd = &cli.Command{
+	Name:      "query-ask",
+	Usage:     "Find a miners ask",
+	ArgsUsage: "[minerAddress]",
```
Comment (dirkmc) [inline cmd/boost/provider_cmd.go:116]: In the libp2p-info command we call the argument "provider address". Happy to use either but suggest we change them all to be consistent.

File: cmd/boost/provider_cmd.go [hunk 3]
```diff
@@ -97,3 +109,192 @@ var libp2pInfoCmd = &cli.Command{
 		return nil
 	},
 }
+
+var queryAskCmd = &cli.Command{
+	Name:      "query-ask",
+	Usage:     "Find a miners ask",
```
Comment (dirkmc) [inline cmd/boost/provider_cmd.go:115]:
```suggestion
	Usage:     "Query a storage provider's storage ask",
```

File: cmd/boost/provider_cmd.go [hunk 4]
```diff
@@ -97,3 +109,192 @@ var libp2pInfoCmd = &cli.Command{
...
+			Name:  "peerid",
+			Usage: "specify peer ID of node to make query against",
+		},
+		&cli.Int64Flag{
+			Name:  "size",
+			Usage: "data size in bytes",
+		},
+		&cli.Int64Flag{
+			Name:  "duration",
+			Usage: "deal duration",
```
Comment (dirkmc) [inline cmd/boost/provider_cmd.go:128]:
```suggestion
			Usage: "deal duration in epochs",
```

File: cmd/boost/provider_cmd.go [hunk 5]
```diff
@@ -97,3 +109,192 @@ var libp2pInfoCmd = &cli.Command{
 		return nil
 	},
 }
+
+var queryAskCmd = &cli.Command{
+	Name:      "query-ask",
```
Comment (dirkmc) [inline cmd/boost/provider_cmd.go:114]: Suggest we call the commands "storage-ask" and "retrieval-ask"

File: cmd/boost/provider_cmd.go [hunk 6]
```diff
@@ -97,3 +109,192 @@ var libp2pInfoCmd = &cli.Command{
...
+		}
+		afmt.Printf("Total Price: %s\n", types.FIL(types.BigMul(perEpoch, types.NewInt(uint64(duration)))))
+
+		return nil
+	},
+}
+
+var queryRetrievalAskCmd = &cli.Command{
+	Name:      "retrieval-ask",
+	Usage:     "Get a miner's retrieval ask",
```
Comment (dirkmc) [inline cmd/boost/provider_cmd.go:210]:
```suggestion
	Usage:     "Query a storage provider's retrieval ask",
```

File: cmd/boost/provider_cmd.go [hunk 7]
```diff
@@ -97,3 +109,192 @@ var libp2pInfoCmd = &cli.Command{
...
+
+		if err := doRpc(ctx, s, &req, &ask); err != nil {
+			return fmt.Errorf("send retrieval-ask request rpc: %w", err)
+		}
+
+		afmt.Printf("Status: %d\n", ask.Status)
+		if ask.Status != 0 {
+			return nil
+		}
+		afmt.Printf("Ask: %s\n", maddr)
```
Comment (dirkmc) [inline cmd/boost/provider_cmd.go:280]: Suggest option for json output here too

File: cmd/boost/provider_cmd.go [hunk 8]
```diff
@@ -97,3 +109,192 @@ var libp2pInfoCmd = &cli.Command{
...
+			Miner: maddr,
+		}
+
+		if err := doRpc(ctx, s, &askRequest, &resp); err != nil {
+			return fmt.Errorf("send ask request rpc: %w", err)
+		}
+
+		ask := resp.Ask.Ask
+
+		afmt.Printf("Ask: %s\n", maddr)
```
Comment (dirkmc) [inline cmd/boost/provider_cmd.go:185]: Suggest we also parse the "json" flag and if it's true, output the results as json (see line 91)

File: cmd/boost/provider_cmd.go [hunk 9]
```diff
@@ -97,3 +109,192 @@ var libp2pInfoCmd = &cli.Command{
...
+
+var queryAskCmd = &cli.Command{
+	Name:      "query-ask",
+	Usage:     "Find a miners ask",
+	ArgsUsage: "[minerAddress]",
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:  "peerid",
+			Usage: "specify peer ID of node to make query against",
+		},
```
Comment (dirkmc) [inline cmd/boost/provider_cmd.go:121]: It seems like the peerid flag is ignored in the code, and we don't have a peerid flag for retrieval ask, so I suggest we just delete it

**Review by dirkmc — APPROVED** [review]
Suggested a few minor changes but overall LGTM 👍

File: cmd/boost/provider_cmd.go (hunk 2 above)
Comment (nonsense) [reply cmd/boost/provider_cmd.go:116]: I was going to raise that up! I copied the commands from Lotus, but yeah, I also think we should be consistent.. Changing this now.

**Comment by dirkmc** [comment]
For convenience I think it makes sense for these commands to take the provider address as a command line parameter rather than a flag:
```
storage-ask <provider addr>
rather than
storage-ask --provider=<provider addr>
```

**Comment by nonsense** [comment]
@dirkmc I'll merge as is and tackle the JSON output across all commands in Boost, not just these two.

**Pull request merged**